actionpack/lib/action_dispatch/middleware/server_timing.rb

```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+module ActionDispatch
+  class ServerTiming
+    class << self
+      def add_timing(key:, dur: nil, desc: nil)
+        timings[key] = { dur: dur, desc: desc }
+      end
+
+      def to_header
+        headers = []
+        timings.each_pair do |key, data|
+          header = "#{key}"
+          header += ";desc=\"#{data[:desc]}\"" if data[:desc]
+          header += ";dur=#{data[:dur]}" if data[:dur]
+          headers << header
+        end
+        headers.join(", ")
+      end
+
+      def measure(key:, desc: nil, &blk)
+        ActiveSupport::Notifications.instrument("measure.server_timing", desc: desc, key: key) do
+          yield
+        end
+      end
+
+      def clear_timings!
+        Thread.current[:server_timing_registry] = {}
+      end
+
+      def timings
+        Thread.current[:server_timing_registry] ||= {}
+      end
+    end
+
+    def initialize(app, default_events = [])
+      @app = app
+      subscribe_to_events(default_events) unless default_events.empty?
+    end
+
+    def call(env)
+      @app.call(env).tap do |response|
```

Comment:
Small nitpick, it'd be more readable to yield the 3 arguments from the response array so you don't have to do `response[1]` ` @app.call(env).tap do |_, headers, _|`